ignite/metrics/entropy.py

```diff
@@ -0,0 +1,82 @@
+
+__all__ = ["Entropy"]
+
+
+class Entropy(Metric):
+    r"""Calculates the mean of `entropy <https://en.wikipedia.org/wiki/Entropy_(information_theory)>`_.
+
+    .. math:: H = \frac{1}{N} \sum_{i=1}^N \sum_{c=1}^C -p_{i,c} \log p_{i,c},
+       \quad p_{i,c} = \frac{\exp(z_{i,c})}{\sum_{c'=1}^C \exp(z_{i,c'})}
+
+    where :math:`p_{i,c}` is the prediction probability of :math:`i`-th data belonging to the class :math:`c`.
+
+    - ``update`` must receive output of the form ``(y_pred, y)`` while ``y`` is not used in this metric.
+    - ``y_pred`` is expected to be the unnormalized logits for each class.
+
+    Args:
+        output_transform: a callable that is used to transform the
+            :class:`~ignite.engine.engine.Engine`'s ``process_function``'s output into the
+            form expected by the metric. This can be useful if, for example, you have a multi-output model and
+            you want to compute the metric with respect to one of the outputs.
+            By default, metrics require the output as ``(y_pred, y)`` or ``{'y_pred': y_pred, 'y': y}``.
+        device: specifies which device updates are accumulated on. Setting the
+            metric's device to be the same as your ``update`` arguments ensures the ``update`` method is
+            non-blocking. By default, CPU.
+
+    Examples:
+        To use with ``Engine`` and ``process_function``, simply attach the metric instance to the engine.
+        The output of the engine's ``process_function`` needs to be in the format of
+        ``(y_pred, y)`` or ``{'y_pred': y_pred, 'y': y, ...}``. If not, ``output_tranform`` can be added
+        to the metric to transform the output into the form expected by the metric.
+
+        For more information on how metric works with :class:`~ignite.engine.engine.Engine`, visit :ref:`attach-engine`.
+
+        .. include:: defaults.rst
+            :start-after: :orphan:
+
+        .. testcode::
+
+            metric = Entropy()
+            metric.attach(default_evaluator, 'entropy')
+            y_true = torch.tensor([0, 1, 2])  # not considered in the Entropy metric.
+            y_pred = torch.tensor([
+                [ 0.0000,  0.6931,  1.0986],
+                [ 1.3863,  1.6094,  1.6094],
+                [ 0.0000, -2.3026, -2.3026]
+            ])
+            state = default_evaluator.run([[y_pred, y_true]])
+            print(state.metrics['entropy'])
+
+        .. testoutput::
+
+            0.8902875582377116
+    """
+
+    _state_dict_all_req_keys = ("_sum_of_entropies", "_num_examples")
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self._sum_of_entropies = torch.tensor(0.0, device=self._device)
+        self._num_examples = 0
+
+    @reinit__is_reduced
+    def update(self, output: Sequence[torch.Tensor]) -> None:
+        y_pred = output[0].detach()
+        prob = F.softmax(y_pred, dim=1)
+        log_prob = F.log_softmax(y_pred, dim=1)
```

Comment:
I think as `y_pred` is already a logit we may not need to reapply log_softmax ?

Reply:
Thank you for your review.
log_softmax is necessary as $\log p_c = \log \text{softmax}(y_\text{pred}) =  y_{\text{pred},c} - \log \sum_i \exp (y_{\text{pred},i}) \neq y_{\text{pred},c}$

> Maybe, we could also reuse torch built-in cross entropy function and provide softmaxed y_pred as target?

Could you please explain in more detail?

Reply:
Yes, you are right there is additional log sum term.

> Could you please explain in more detail?

I was thinking that cross entropy mplementation could fit what we want to compute here. In details, actually, it is not exactly usable for our purpose as cross-entropy takes log_p at places when target == 1.

https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss